README.md
lmds
================

<!-- badges: start -->
[![R-CMD-check](https://github.com/dynverse/lmds/actions/workflows/R-CMD-check.yaml/badge.svg)](https://github.com/dynverse/lmds/actions/workflows/R-CMD-check.yaml)
<!-- badges: end -->

# `lmds`: Landmark Multi-Dimensional Scaling

A fast dimensionality reduction method scaleable to large numbers of
samples. Landmark Multi-Dimensional Scaling (LMDS) is an extension of
classical Torgerson MDS, but rather than calculating a complete distance
matrix between all pairs of samples, only the distances between a set of
landmarks and the samples are calculated.

``` r
library(lmds)
x <- as.matrix(iris[,1:4])
dimred <- lmds(x, ndim = 2)
qplot(dimred[,1], dimred[,2]) + labs(title = "lmds()") + theme_classic()
#> Warning: `qplot()` was deprecated in ggplot2 3.4.0.
#> This warning is displayed once every 8 hours.
#> Call `lifecycle::last_lifecycle_warnings()` to see where this warning was generated.
```

<img src="man/figures/compare-1.png" width="100%" />

``` r
dimred <- cmdscale(dist(x))
qplot(dimred[,1], dimred[,2]) + labs(title = "cmdscale()") + theme_classic()
```

<img src="man/figures/compare-2.png" width="100%" />

## Execution time

The execution time of `lmds()` scales linearly with respect to the
dataset size. <img src="man/figures/timings-1.png" width="100%" />

## Latest changes

Check out `news(package = "lmds")` or [NEWS.md](NEWS.md) for a full list
of changes.

<!-- This section gets automatically generated from inst/NEWS.md, and also generates inst/NEWS -->

### Recent changes in lmds 0.1.0

Initial release of lmds.

-   A fast dimensionality reduction method scaleable to large numbers of
    samples. Landmark Multi-Dimensional Scaling (LMDS) is an extension
    of classical Torgerson MDS, but rather than calculating a complete
    distance matrix between all pairs of samples, only the distances
    between a set of landmarks and the samples are calculated.


NEWS.md
# lmds 0.1.0

Initial release of lmds.

* A fast dimensionality reduction method scaleable to large numbers of samples.
  Landmark Multi-Dimensional Scaling (LMDS) is an extension of classical Torgerson MDS,
  but rather than calculating a complete distance matrix between all pairs of samples,
  only the distances between a set of landmarks and the samples are calculated.


tests/testthat.R
library(testthat)
library(lmds)

test_check("lmds")


tests/testthat/test-lmds.R
context("Testing lmds()")

set.seed(1)
x <- as.matrix(iris[,1:4])

test_that("Perform dimred", {
  out1 <- lmds(x)

  expect_is(out1, "matrix")
  expect_identical(rownames(out1), rownames(x))
  expect_identical(colnames(out1), paste0("comp_", seq_len(ncol(out1))))

  set.seed(1)
  out2 <- lmds(x, ndim = 2)
  expect_is(out2, "matrix")
  expect_identical(rownames(out2), rownames(x))
  expect_identical(colnames(out2), paste0("comp_", seq_len(ncol(out2))))
  expect_equal(ncol(out2), 2)

  v1 <- as.vector(as.matrix(dist(out1)))
  v2 <- as.vector(as.matrix(dist(out2)))

  expect_true(all(cor(cbind(v1, v2)) > .25))
})

largex <- Matrix::rsparsematrix(10000, 1000, .05)

test_that("P^erform dimred on large matrix", {
  out1 <- lmds(largex)
  expect_is(out1, "matrix")
  expect_identical(rownames(out1), rownames(largex))
  expect_identical(colnames(out1), paste0("comp_", seq_len(ncol(out1))))
})


tests/testthat/test-select_landmarks.R
context("Testing select_landmarks()")

x <- as.matrix(iris[,1:4])
test_that("Selecting landmarks", {
  lms <- select_landmarks(x, distance_method = "euclidean", landmark_method = "sample", num_landmarks = 500)

  expect_is(lms, "matrix")
  expect_gte(min(lms), 0)
  expect_equal(dim(lms), c(150, nrow(x)))

  lix <- attr(lms, "landmark_ix")
  expect_false(is.null(lix))
  expect_true(all(lix >= 1 & lix <= nrow(x)))

  expect_equivalent(
    lms[cbind(seq_along(lix), lix)],
    rep(0, length(lix)),
    tolerance = 1e-4
  )

  dis <- as.matrix(dist(x))[lix, , drop = FALSE]
  expect_equal(as.vector(dis), as.vector(lms), tolerance = 1e-4)
})

test_that("Selecting landmarks with different parameters", {
  lms <- select_landmarks(x, distance_method = "pearson", landmark_method = "sample", num_landmarks = 101)

  expect_is(lms, "matrix")
  expect_gte(min(lms), 0)
  expect_equal(dim(lms), c(101, nrow(x)))

  lix <- attr(lms, "landmark_ix")
  expect_false(is.null(lix))

  expect_true(all(lix >= 1 & lix <= nrow(x)))

  expect_equivalent(
    lms[cbind(seq_along(lix), lix)],
    rep(0, length(lix)),
    tolerance = 1e-4
  )

  dis <- (1 - (cor(t(x)) + 1) / 2)[lix, , drop = FALSE]
  expect_equal(as.vector(dis), as.vector(lms), tolerance = 1e-4)
})


R/lmds.R
#' Landmark MDS
#'
#' A fast dimensionality reduction method scaleable to large numbers of samples.
#' Landmark Multi-Dimensional Scaling (LMDS) is an extension of classical 'Torgerson MDS',
#' but rather than calculating a complete distance matrix between all pairs of samples,
#' only the distances between a set of landmarks and the samples are calculated.
#'
#' @inheritParams select_landmarks
#' @inheritParams cmdscale_landmarks
#'
#' @export
#'
#' @include cmdscale_landmarks.R select_landmarks.R
#'
#' @return The dimensionality reduction in the form of a `nrow(x)` by `ndim` matrix.
#'
#' @examples
#' library(Matrix)
#' x <- Matrix::rsparsematrix(1000, 1000, .01)
#' lmds(x, ndim = 3)
lmds <- dynutils::inherit_default_params(
  list(select_landmarks, cmdscale_landmarks),
  function(
    x,
    ndim,
    distance_method,
    landmark_method,
    num_landmarks
  ) {
    # select the landmarks
    dist_2lm <- select_landmarks(
      x = x,
      distance_method = distance_method,
      landmark_method = landmark_method,
      num_landmarks = num_landmarks
    )

    # reduce dimensionality for landmarks and project to non-landmarks
    dimred <- cmdscale_landmarks(
      dist_2lm = dist_2lm,
      ndim = ndim,
      rescale = TRUE
    )

    rownames(dimred) <- rownames(x)


    dimred
  }
)


R/cmdscale_landmarks.R
#' Perform MDS on landmarks and project other samples to the same space
#'
#' @param dist_2lm Distance matrix between the landmarks and all the samples in original dataset
#' @param ndim The number of dimensions
#' @param rescale Whether or not to rescale the final dimensionality reduction (recommended)
#' @param ... Extra params to pass to [irlba::irlba()]
#'
#' @importFrom irlba partial_eigen
#' @importFrom dynutils scale_uniform
#'
#' @return The dimensionality reduction in the form of a `ncol(dist_2lm)` by `ndim` matrix.
#'
#' @export
#'
#' @examples
#' library(Matrix)
#' x <- as.matrix(iris[,1:4])
#' dist_2lm <- select_landmarks(x)
#' cmdscale_landmarks(dist_2lm)
cmdscale_landmarks <- function(dist_2lm, ndim = 3, rescale = TRUE, ...) {
  assert_that(
    !is.null(attr(dist_2lm, "landmark_ix")) || (!is.null(rownames(dist_2lm)) && !is.null(colnames(dist_2lm))),
    is.numeric(ndim), length(ndim) == 1, ndim >= 1,
    is.logical(rescale), length(rescale) == 1, !is.na(rescale)
  )

  ix_lm <- attr(dist_2lm, "landmark_ix")
  if (is.null(ix_lm)) {
    ix_lm <- match(rownames(dist_2lm), colnames(dist_2lm))
  }

  # short hand notations
  x <- dist_2lm[, ix_lm, drop = FALSE]^2
  n <- as.integer(nrow(x))
  N <- as.integer(ncol(dist_2lm))

  # double center data
  mu_n <- rowMeans(x)
  mu <- mean(x)
  x_c <- sweep(x, 1, mu_n, "-")
  x_dc <- sweep(x_c, 2, mu_n, "-") + mu

  # classical MDS on landmarks
  if (ndim > 0.5 * min(nrow(x_dc), ncol(x_dc))) {
    e <- eigen(-x_dc / 2, symmetric = TRUE)
    e$values <- e$values[seq_len(ndim)]
    e$vectors <- e$vectors[, seq_len(ndim), drop = FALSE]
  } else {
    e <- irlba::partial_eigen(-x_dc / 2, symmetric = TRUE, n = ndim, ...)
  }
  ev <- e$values
  evec <- e$vectors
  ndim1 <- sum(ev > 0)
  if (ndim1 < ndim) {
    warning(gettextf("only %d of the first %d eigenvalues are > 0", ndim1, ndim), domain = NA)
    evec <- evec[, ev > 0, drop = FALSE]
    ev <- ev[ev > 0]
    ndim <- ndim1
  }

  # distance-based triangulation
  points_inv <- evec / rep(sqrt(ev), each = n)
  dimred <- -t(t(points_inv) %*% ((dist_2lm - rep(mu_n, each = N)) / 2))
  dimred <- (-t(dist_2lm - rep(mu_n, each = N)) / 2) %*% points_inv

  if (rescale) {
    dimred <- dynutils::scale_uniform(dimred)
  }

  colnames(dimred) <- paste0("comp_", seq_len(ndim))

  dimred
}


R/select_landmarks.R
#' Select landmarks from dataset
#'
#' In addition, the distances between the landmarks and all samples are calculated.
#'
#' @param x A matrix, optionally sparse.
#' @param distance_method The distance metric to use. Options are "euclidean" (default), "pearson", "spearman", "cosine", "manhattan".
#' @param landmark_method The landmark selection method to use. Options are "sample" (default).
#' @param num_landmarks The number of landmarks to use,
#'
#' @return The distance matrix between the landmarks and all samples. In addition, an attribute `"landmark_ix"`
#'   denotes the indices of landmarks that were sampled.
#'
#' @export
#'
#' @examples
#' library(Matrix)
#' x <- Matrix::rsparsematrix(1000, 1000, .01)
#' select_landmarks(x)
select_landmarks <- function(
  x,
  distance_method,
  landmark_method = c("sample"),
  num_landmarks = 500
) {
  distance_method <- match.arg(distance_method)
  landmark_method <- match.arg(landmark_method)
  assert_that(
    is.matrix(x) || is_sparse(x),
    is.numeric(num_landmarks), length(num_landmarks) == 1, num_landmarks > 2
  )

  # parameter check on num_landmarks
  if (num_landmarks > nrow(x)) {
    num_landmarks <- nrow(x)
  }

  # naive -> just subsample the cell ids
  if (landmark_method == "sample") {
    ix_lm <- sample.int(nrow(x), num_landmarks)
    dist_2lm <- as.matrix(calculate_distance(x[ix_lm, , drop = FALSE], x, method = distance_method))
  }

  dist_2lm <- zapsmall(dist_2lm)

  attr(dist_2lm, "landmark_ix") <- ix_lm

  dist_2lm
}
formals(select_landmarks)$distance_method <- unique(c("euclidean", dynutils::list_distance_methods())) # prioritise euclidean
